# Week8. 

# CNN_assignment_AlexNet_parameter 
13기 이유민

---

---

### <계산에 사용한 식>
- $OH(convolution) = \frac{IH + 2P - FH}{S} +1$  
- $OH(pooling) = \frac{IH - FH}{S} +1$  

### <Alexnet 특징>
1. Fully Connected layer 3개와 Convolutional Layer어 5개로 구성
2. Activation Function이 Relu -> 학습시간 단축!
5. 256x256이미지를 랜덤으로 227x227로 학습데이터를 늘려 overfitting 방지
---

# AlexNet

- [paper](http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf)

- [imagenet data(2012)](http://image-net.org/challenges/LSVRC/2012/index#task)

- [code](https://pytorch.org/docs/0.4.0/_modules/torchvision/models/alexnet.html)

- Model architecture
![model_architecture](https://cv-tricks.com/wp-content/uploads/2017/03/xalexnet_small-1.png.pagespeed.ic.u_mv-jhXMI.webp)

### Naive Version
CONV_1 - POOL_1 - CONV_2 - POOL_2 - CONV_3 - CONV_4 - CONV_5 - POOL_3 - FC1 - FC2 - FC3 (->SOFTMAX)

### detailed
CONV_1(ReLU) - POOL_1 - CONV_2(ReLU) - POOL_2 - CONV_3(ReLU) - CONV_4(ReLU) - CONV_5(ReLU) - POOL_3 -(Flatten) FC1(ReLU) - FC2(ReLU) - FC3(->SOFTMAX)

---

# Q.1 fill in the '????' & 'TODO'

### Layer 1 is a Convolution Layer_1

- **Input Image size**     224 x 224 x 3 -> 227 x 227 x 3 (표기 실수로 인한 수정)
  - IH:227 x IW:227 x IC:3

- **Number of filters**   96
  - FN:96

- **Filter size** 11 x 11 x 3
  - FH:11 x FW:11 x FC(항상 IC랑 같다):3

- **Stride** 4
  - S:4

- **Layer 1 Output**  55 x 55 x 96 (because of stride 4)  OH x OW x FN
  - OH:55 x OW:55 x FN(위의 FN이랑 같다):96 
  - 아래 계산과정 참고
  - 이게 다시 Input이 된다. OH -> IH, OW -> IW, FN -> IC


In [1]:
# output size = (IH + 2*P - FH) / S + 1
(227 + 2*0 - 11) / 4 + 1 

55.0

output의 volume이 55.0로 정수값을 가지네요!

따라서 filter size가 input 데이터를 벗어나지 않음을 알 수 있습니다

In [2]:
## TODO ##

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

# (FH * FW * FC) * FN + B 의 값인
Conv_1 = (11 * 11 * 3) * 96 + 96 
'''last 96 for Bias'''

'last 96 for Bias'

In [3]:
Conv_1 # 파라미터 수

34944

### Layer 2 is a Max Pooling_1 Followed by Convolution_1

- **Input**  55 x 55 x 96
  - IH:55 x IW:55 x IC:96
- **Max pooling**  

- **Pooling size**(overlapping) 3 x 3
  - FH:3 x FW:3

- **Stride** 2
  - S:2

- **Layer 2 Output** 27 x 27 x 96

In [4]:
# output size = (IH - FH) / S + 1 = 
(55 - 3) / 2 + 1 

27.0

In [5]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Max_pool_1 = 0

In [6]:
Max_pool_1
# Maxpooling에서는 학습되는 파라미터가 없기 때문입니다

0

+) 1st, 2nd, 5th Convolutional Layer에 적용하는 Maxpooling은 오버피팅을 막는 데 도움을 줍니다.

(pooling size =3 stride =2로 겹치게 해서 사용)

### Layer 3 is a a Convolution Layer_2

- **Input**  27 x 27 x 96

- **Number of filters**  256

- **Filter size**  5 x 5 x 96 

- **Stride** 1

- **padding** 2

- **Layer 3 Output** 27 x 27 x 256


In [7]:
# output size = (IH + 2*P - FH) / S + 1
(27 + 2*2 - 5) / 1 + 1

27.0

In [8]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Conv_2 = (5 * 5 * 96) * 256 + 256

In [9]:
Conv_2

614656

### Layer 4 is a Max Pooling_2 Followed by Convolution_2

- **Input**  27 x 27 x 256

- **Max pooling**  

- **Pooling size**(overlapping) 3 x 3

- **Stride** 2

- **Layer 4 Output**  13 x 13 x 256

In [10]:
# output size = (IH - FH) / S + 1 
(27 - 3) / 2 + 1

13.0

In [11]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Max_pool_2 = 0

In [12]:
Max_pool_2

0

### Layer 5 is a a Convolution Layer_3

- **Input**  13 x 13 x 256

- **Number of filters**  384

- **Filter size**  3 x 3 x 256

- **Stride** 1

- **padding** 1

- **Layer 5 Output** 13 x 13 x 384


In [13]:
# output size = (IH + 2*P - FH) / S + 1
(13 + 2*1 - 3) / 1 + 1

13.0

In [14]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Conv_3 = (3 * 3 * 256) * 384 + 384

In [15]:
Conv_3

885120

### Layer 6 is  a Convolution Layer_4

- **Input**  13 x 13 x 384

- **Number of filters**  384

- **Filter size**  3 x 3 x 384

- **Stride** 1

- **padding** 1

- **Layer 6 Output** 13 x 13 x 384

In [16]:
# output size = (IH + 2*P - FH) / S + 1
(13 + 2*1 - 3) / 1 + 1

13.0

In [17]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Conv_4 = (3 * 3 * 384) * 384 + 384

In [18]:
Conv_4

1327488

### Layer 7 is a Convolution Layer_5

- **Input**  13 x 13 x 384

- **Number of filters**  256

- **Filter size**  3 x 3 x 384

- **Stride** 1

- **padding** 1

- **Layer 7 Output** 13 x 13 x 256

In [19]:
# output size = (IH + 2*P - FH) / S + 1
(13 + 2*1 - 3) / 1 + 1

13.0

In [20]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################
Conv_5 =  (3 * 3 * 384) * 256 + 256

In [21]:
Conv_5

884992

### Layer 8 is a Max Pooling_3 Followed by Convolution_5

- **Input**  13 x 13 x 256

- **Max pooling**  

- **Pooling size**(overlapping) 3 x 3

- **Stride** 2

- **Layer 8 Output** 6 x 6 x 256

In [22]:
# output size = (IH - FH) / S + 1
(13 - 3) / 2 + 1

6.0

In [23]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

Max_pool_3 = 0

### Layer 9 is a Fully_Connected layer_1

- **input** 6 x 6 x 256 = 9216

- **flatten** 9216 (= 6 x 6 x 256)

- **output size** (N,flatten) x (flatten,4096)
  - 여기서 4096 : 하이퍼 파라미터 (계산한 값이 아님)

- **N** Number of input data = 6 x 6 x 256

In [24]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################
FC1 =  (6 * 6 * 256) * 4096 + 4096

In [25]:
FC1

37752832

### Layer 10 is a Fully_Connected layer_2

- **input** (N,4096)

- **output size** (N,4096) x (4096,4096)

- **N** Number of input data = 6

In [26]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

FC2 = (4096 * 4096 * 1) + 4096

In [27]:
FC2

16781312

### Layer 11 is a Fully_Connected layer_3

- **input** (N,4096)

- **output size** (N,4096) x (4096,Num_classes)

- **N** Number of input data

- **Num_classes** Number of labels + 4096 # +4096 의미?

In [28]:
## TODO 

######################################################
#  Calculate the number of parameters in this layer  #
######################################################

FC3 = 4096 * 1000 + 1000

In [29]:
FC3

4097000

:: softmax를 통해 1000개의 카테고리로 분류

---
## [ 추가로 학습한 부분 ]

### 1. Alexnet의 parameter 수

In [30]:
# parameter 수
Conv_1 + Max_pool_1 + Conv_2 + Max_pool_2 + Conv_3 + Conv_4 + Conv_5 + Max_pool_3 + FC1 + FC2 + FC3

62378344

따라서 AlexNet의 전체 parameter의 합은 62378344개 !

### 2. Alexnet의 성능을 향상시키려면 ?

1. GPU를 2개 사용
2. ReLU 
    - Alexnet의 망은 크기가 크기 때문에 속도에 영향 많이 받음
    - ReLU를 쓰면 sigmoid 함수를 쓸 때보다 학습속도가 약 6배 정도 증가함
3. Local Response Normalization
    1. 같은 위치에서의 response를 커널들의 결과와 비교해서 강한 자극만 전달
    2. ReLU의 경우 input의 normaization 불필요
    3. max pooling 전에 사용하면 됨
4. Overfitting 해결
    1. data augmentation
    2. dropout